### ISSUE TITLE
[Truffle] Issue passing arg in test runner

### ISSUE THREAD

**Issue by bjfish** [jruby/jruby#2963]
I had a hard time to create a minimal case for this but it should be easy to recreate with `jt test ruby/test_pipe.rb` which gives the following errors here and in a number of other MRI tests. The `NilClass` argument should be the argument `:empty?`.

```
  1) Error:
TestPipe#test_eof_0_rw:
TypeError: Truffle doesn't have a case for the org.jruby.truffle.nodes.core.KernelNodesFactory$RespondToNodeFactory$RespondToNodeGen node with values of type  Array(org.jruby.truffle.runtime.core.RubyArray)[null,0] NilClass(org.jruby.truffle.runtime.core.RubyBasicObject) org.jruby.truffle.runtime.UndefinedPlaceholder
    	from /Users/brandonfish/Documents/jruby/test/mri/lib/test/unit/assertions.rb:364:in `assert_respond_to'
    	from test/mri/runner.rb:25:in `after_teardown'
    	from /Users/brandonfish/Documents/jruby/test/mri/lib/test/unit/testcase.rb:17:in `run'
    	from /Users/brandonfish/Documents/jruby/test/mri/lib/test/unit.rb:682:in `block in _run_suites'
    	from /Users/brandonfish/Documents/jruby/test/mri/lib/test/unit.rb:680:in `_run_suites'
    	from /Users/brandonfish/Documents/jruby/test/mri/lib/test/unit.rb:31:in `run'
    	from /Users/brandonfish/Documents/jruby/test/mri/lib/test/unit.rb:799:in `run'
    	from /Users/brandonfish/Documents/jruby/test/mri/lib/test/unit.rb:859:in `run'
    	from /Users/brandonfish/Documents/jruby/test/mri/lib/test/unit.rb:863:in `run'
    	from test/mri/runner.rb:41:in `<main>'

  2) Error:
TestPipe::WithConversion#test_eof_0_rw:
TypeError: Truffle doesn't have a case for the org.jruby.truffle.nodes.core.KernelNodesFactory$RespondToNodeFactory$RespondToNodeGen node with values of type  Array(org.jruby.truffle.runtime.core.RubyArray)[null,0] NilClass(org.jruby.truffle.runtime.core.RubyBasicObject) org.jruby.truffle.runtime.UndefinedPlaceholder
    	from /Users/brandonfish/Documents/jruby/test/mri/lib/test/unit/assertions.rb:364:in `assert_respond_to'
    	from test/mri/runner.rb:25:in `after_teardown'
    	from /Users/brandonfish/Documents/jruby/test/mri/lib/test/unit/testcase.rb:17:in `run'
    	from /Users/brandonfish/Documents/jruby/test/mri/lib/test/unit.rb:682:in `block in _run_suites'
    	from /Users/brandonfish/Documents/jruby/test/mri/lib/test/unit.rb:680:in `_run_suites'
    	from /Users/brandonfish/Documents/jruby/test/mri/lib/test/unit.rb:31:in `run'
    	from /Users/brandonfish/Documents/jruby/test/mri/lib/test/unit.rb:799:in `run'
    	from /Users/brandonfish/Documents/jruby/test/mri/lib/test/unit.rb:859:in `run'
    	from /Users/brandonfish/Documents/jruby/test/mri/lib/test/unit.rb:863:in `run'
    	from test/mri/runner.rb:41:in `<main>'

```

**Comment by eregon** [comment]
Yeah I noticed it recently as well on all mri tests I tried.
I think doing a bisect here could help to know what went wrong.

**Comment by bjfish** [comment]
@eregon @chrisseaton It looks like it started happening at this commit:
https://github.com/jruby/jruby/commit/79eee3f69dabe438be366b951ffe8fc857516266

All the MRI tests started to fail:
https://travis-ci.org/jruby/jruby/jobs/63194822

**Comment by eregon** [comment]
Thanks for looking into it! I'll assign Chris then.

**Comment by bjfish** [comment]
It looks like there is another new separate issue when `jt test mri` is run which results in the following error:
```
$ jt test mri
core:/core/rubinius/common/type.rb:62:in `coerce_to_failed': Coercion error: nil.to_str => String failed (TypeError)
	from core:/core/rubinius/common/type.rb:52:in `execute_coerce_to'
	from core:/core/rubinius/common/type.rb:45:in `coerce_to'
	from core:/core/rubinius/common/kernel.rb:155:in `StringValue'
	from core:/core/rubinius/common/type.rb:419:in `coerce_to_path'
	from core:/core/rubinius/common/file.rb:381:in `dirname'
	from /Users/brandonfish/Documents/jruby/test/mri/ruby/envutil.rb:558:in `RbConfig'
	from /Users/brandonfish/Documents/jruby/test/mri/ruby/envutil.rb:552:in `<main>'
	from test/mri/runner.rb:8:in `<main>'
```
This was the first build with this error: https://travis-ci.org/jruby/jruby/builds/63422709

So it is probably one of the following commits for the new issue:
7a9d17cddba033709f2eee8be76632aa9c4397a6
f25eec841e68cf1cdc4241f4f4873974e1936f8a
af92b097e7ee75faa87c279bfe7167a61882499b
ded007da4b31e35cdf00d7b0af3101cbbdb58e71
a27a4692a98e2726a3d8249c5207e8aebe23b70d

**Comment by chrisseaton** [comment]
`jt test ruby/test_pipe.rb` works for me at the moment - but with `0 files, 0 examples, 0 expectations, 0 failures, 0 errors, 0 tagged` - can you confirm or close?

**Comment by eregon** [comment]
This is due to a change of jt to run specs by default. Try with the full path to the test.

**Comment by eregon** [comment]
(Or `jt test mri ...`)

**Comment by chrisseaton** [comment]
I see it now.

**Comment by chrisseaton** [comment]
This stack trace makes no sense to me: we're on a line with a call to `assert_empty` and then we're in `assert_respond_to`. Does anyone understand the logic here?

```
    	from /Users/chrisseaton/Documents/ruby/jruby/test/mri/lib/test/unit/assertions.rb:362:in `assert_respond_to'
    	from test/mri/runner.rb:25:in `after_teardown'
``

**Comment by bjfish** [comment]
@chrisseaton  It looks like the stack trace is missing a line, it should probably look like this:
```
       from /Users/chrisseaton/Documents/ruby/jruby/test/mri/lib/test/unit/assertions.rb:362:in `assert_respond_to'
       from test/mri/lib/minitest/unit.rb:216:in `assert_empty' 
       from test/mri/runner.rb:25:in `after_teardown'
```

**Comment by chrisseaton** [comment]
Test case:

```
class A
  def foo(a, b, c)
    p [a, b, c]
  end
end

class B < A
  def foo(a, (b, c), d = nil)
    super
  end
end

B.new.foo(:a, :b)
```

**Comment by chrisseaton** [comment]
This broke when I fixed other things that were more badly broken in super calls.

**Comment by bjfish** [comment]
@chrisseaton Yes, i think it's one of the four commits mentioned here based on what I saw in travis: https://github.com/jruby/jruby/issues/2963#issuecomment-104045437

**Comment by chrisseaton** [comment]
Yeah but I mean the fix won't be found by looking at those commits - but I think I know how to fix it now.

**Comment by chrisseaton** [comment]
Fixed in 209bc659d74155072c1dc90cda22d6b8a8c43182

**Comment by bjfish** [comment]
There is a little regression in the mri tests.
```
  1) Failure:
TestSuper#test_array2 [	from /Users/brandonfish/Documents/jruby/test/mri/ruby/test_super.rb:88]:
Failed assertion, no message given.

  2) Failure:
TestSuper#test_array3 [	from /Users/brandonfish/Documents/jruby/test/mri/ruby/test_super.rb:92]:
Failed assertion, no message given.

  3) Failure:
TestSuper#test_array4 [	from /Users/brandonfish/Documents/jruby/test/mri/ruby/test_super.rb:96]:
Failed assertion, no message given.

  4) Error:
TestSuper#test_double2:
ArgumentError: wrong number of arguments (1 for 2)
    /Users/brandonfish/Documents/jruby/test/mri/ruby/test_super.rb:7:in `double': wrong number of arguments (1 for 2) (ArgumentError)
    	from /Users/brandonfish/Documents/jruby/test/mri/ruby/test_super.rb:22:in `double'
    	from /Users/brandonfish/Documents/jruby/test/mri/ruby/test_super.rb:77:in `test_double2'
    	from /Users/brandonfish/Documents/jruby/test/mri/lib/test/unit.rb:888:in `run_test'
```

**Comment by chrisseaton** [comment]
We've got a few failing super specs in RubySpec as well - unless they are blocking something I wasn't going to prioritise them now (even if they were working before).

**Comment by bjfish** [comment]
@eregon @chrisseaton @nirvdrum I'll mention regressions in MRI tests if I see them just for informational purposes. Feel free to fix them or not.

**Comment by eregon** [comment]
Regressions fixed in 911a6f0df690dca788bdd0f43ce9e96983a65e22 (and parents).
Will merge after the release.
